framework/opt/maybe.go

```diff
@@ -0,0 +1,78 @@
+package opt
+
+import (
+	"encoding/json"
+)
+
+// Maybe is a simple implementation of an optional value type.
+type Maybe[V any] struct {
+	defined bool
+	value   V
+}
+
+// Some returns a Maybe that has a defined value.
+func Some[V any](value V) Maybe[V] {
+	return Maybe[V]{defined: true, value: value}
+}
+
+// None returns a Maybe with no value.
+func None[V any]() Maybe[V] { return Maybe[V]{} }
+
+// FromPtr returns a Maybe that has a defined value of *ptr if ptr is non-nil, or
+// no value if ptr is nil.
+func FromPtr[V any](ptr *V) Maybe[V] {
+	if ptr != nil {
+		return Some[V](*ptr)
+	}
+	return None[V]()
+}
+
+// IsDefined returns true if the Maybe has a value.
+func (m Maybe[V]) IsDefined() bool { return m.defined }
```

Comment:
suggestion: since the "defined" variant is called `Some`, this could be `IsSome`.
Could also add `IsNone` for convenience.

Reply:
`IsDefined` is what it's called in a lot of optional types I've used on other platforms, and we also used it in the optional types in `ldvalue`. Personally I find it clearer than `IsSome` in a case where you might be looking at some code that is calling `IsSome`, and you don't happen to already know that Some is the name of a constructor. It's also what EasyJSON expects a method like this to be named, even though that's not relevant here.

Reply:
Ah, ok, makes sense. I was coming from the rust perspective where the variant is called Some, and it has IsSome/IsNone.